packages/typespec-client-generator-core/src/package.ts

```diff
@@ -508,7 +508,33 @@ function getSdkEndpointParameter(
         );
       }
     }
-    optional = Boolean(servers[0].url.length && templateArguments.every((param) => param.optional));
+    if (templateArguments.length === 0) {
+      templateArguments.push({
+        kind: "path",
+        name: "endpoint",
+        isGeneratedName: true,
+        description: "Service host",
+        onClient: true,
+        urlEncode: false,
+        optional: false,
+        serializedName: "endpoint",
+        correspondingMethodParams: [],
+        type: {
+          kind: "string",
+          encode: "string",
+          decorators: [],
+        },
+        clientDefaultValue: servers[0].url,
```

Comment:
isn't that breaking if this has parameters?

Reply:
I only add if the url is not parameterized

Reply:
this feels like it defeat the point though, if someone does this 
```
@server("https://{region}.myservice.com", "", {region: string})
```
I still would want to be able to provide by local test server/some staging url

Reply:
this is I guess my whole point on why this make 0 sense to have as a compiler/http library api

Reply:
Just a comment, that does not affect the discussion here.

The reason why we need `endpoint` in ARM, is that `@server("https://management.azure.com")` is just wrong.
That URL is only valid for Azure Global.
Azure US Gov, Azure Germany, Azure China all uses different endpoint. E.g. Azure Gov is `"https://management.usgovcloudapi.net/"`.

Reply:
> (though this sound a bit arrogant as we are forcing something to dev -- regardless of how they write the typespec, and their only choice otherwise is not to use our emitter) 
I would argue strongly for the opposite, as a spec writer I wrote the server(s) I want my user to use. But I wouldn't want to describe my dev/stagging/test server in the spec but still have the option to do so. If you feel strongly about this behavior you can always provide an option to disable it. This should be an option when creating a client, user can use it or not.

And yeah would argue the same for code generated from OpenAPI this is the same case.


For the Arm specific case however I kinda agree that relying on this functionality might not be enough. The way I see it is an escape hatch which means a client would provide it but it might not be obvious.

Though I would also compare that to the Github API which is about the same scenario where you have your own hosted entreprise github with a different endpoint

```ts
new Octokit() // Github public endpoint
new Octokit({
  baseUrl: "https://myentreprise.hostedgithub.com",
})
```

so I would be fine additionally from this feature updating the arm library to provide either a list of all the cloud(not ideal) or just say that the host is configurable but the default is arm but I think it really depends on what is the experience for this escape hatch. This will also be a BREAKING change to the swagger which all use the public cloud today

Reply:
> you can always provide an option to disable it. This should be an option when creating a client, user can use it or not.

Yeah, agree this may be a better solution. Codegen can refine that later.

PS: dev can override the endpoint anyway, in current SDK, with the common HttpPipelinePolicy (though it looks bad) e.g.
```java
        new ##ClientBuilder()
            .addPolicy((context, policy) -> {
                context.getHttpRequest().setUrl(context.getHttpRequest().getUrl().toString().replace("http://localhost:3000/", "http://localhost:3001/"));
                return policy.process();
            })
            .buildClient()
```

Reply:
yeah that does feel like a little too much here, but if you had 
```
new ClientBuilder().overrideBaseUrl("https://localhost:30001")
```

instead of explicitly allowing it in the constructor then I think that would satisfy